crates/matrix-sdk/src/event_cache/store.rs

```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
     pub fn events(&self) -> impl Iterator<Item = (Position, &Event)> {
         self.chunks.items()
     }
+
+    /// Deduplicate `events` considering all events in `Self::chunks`.
+    ///
+    /// The returned tuple contains (i) the unique events, and (ii) the
+    /// duplicated events (by ID).
+    fn filter_duplicated_events<'a, I>(&'a mut self, events: I) -> (Vec<Event>, Vec<OwnedEventId>)
+    where
+        I: Iterator<Item = Event> + 'a,
+    {
+        let mut duplicated_event_ids = Vec::new();
+
+        let deduplicated_events = self
+            .deduplicator
+            .scan_and_learn(events, self)
+            .filter_map(|decorated_event| match decorated_event {
+                Decoration::Unique(event) => Some(event),
+                Decoration::Duplicated(event) => {
+                    error!(event_id = ?event.event_id(), "Found a duplicated event");
+
+                    duplicated_event_ids.push(event.event_id().expect("The event has no ID"));
+
+                    // Keep the new event!
+                    Some(event)
+                }
+                Decoration::Invalid(event) => {
+                    error!(?event, "Found an invalid event");
+
+                    None
+                }
+            })
+            .collect();
+
+        (deduplicated_events, duplicated_event_ids)
+    }
+}
+
+// Private implementations, implementation specific.
+impl RoomEvents {
+    /// Remove some events from `Self::chunks`.
+    ///
+    /// This method iterates over all event IDs in `event_ids` and removes the
+    /// associated event (if it exists) from `Self::chunks`.
+    ///
+    /// This is used to remove duplicated events, see
+    /// [`Self::filter_duplicated_events`].
+    fn remove_events(&mut self, event_ids: Vec<OwnedEventId>) {
+        for event_id in event_ids {
+            let Some(event_position) = self.revents().find_map(|(position, event)| {
+                (event.event_id().as_ref() == Some(&event_id)).then_some(position)
+            }) else {
+                error!(?event_id, "An event has been detected, but it's position seems unknown");
```

Comment:
The message is slightly too coupled to the deduplication use case IMO, how about something more generic?

```suggestion
                error!(?event_id, "Couldn't find event to remove");
```

Reply:
I agree. I was hesitant keeping this message. Updating it!